File: Source/Aladdin.CAPI.ANSI.Java/src/aladdin/capi/ansi/keyx/ecdh/KeyAgreement.java
package aladdin.capi.ansi.keyx.ecdh;
import aladdin.*; 
import aladdin.asn1.*; 
import aladdin.asn1.iso.*; 
import aladdin.asn1.ansi.x962.*; 
import aladdin.capi.*; 
import aladdin.capi.derive.*; 
import aladdin.capi.ansi.derive.*; 
import aladdin.math.*; 
import java.security.*; 
import java.security.spec.*; 
import java.io.*; 
import java.math.BigInteger;

///////////////////////////////////////////////////////////////////////////
// Формирование общего ключа Elliptic Curve Diffie-Hellman
///////////////////////////////////////////////////////////////////////////
public class KeyAgreement extends aladdin.capi.KeyAgreement
{
    // способ кодирования чисел
    private static final Endian ENDIAN = Endian.BIG_ENDIAN; 

    private final KeyDerive             keyDerive;          // алгоритм наследования ключа
    private final AlgorithmIdentifier   keyWrapParameters;  // параметры алгоритма шифрования ключа
    private final boolean               useCofactor;        // признак дополнительного умножения
    
    // конструктор
    public KeyAgreement(boolean useCofactor, 
        KeyDerive keyDerive, AlgorithmIdentifier keyWrapParameters) 
    { 
        // сохранить переданные параметры
        this.keyDerive = RefObject.addRef(keyDerive); 
            
        // сохранить переданные параметры
        this.useCofactor = useCofactor; this.keyWrapParameters = keyWrapParameters; 
    } 
    // конструктор
    public KeyAgreement(boolean useCofactor, 
        Hash hashAlgorithm, AlgorithmIdentifier keyWrapParameters) 
    { 
        // сохранить переданные параметры
        this.keyDerive = new X963KDF(hashAlgorithm); 
            
        // сохранить переданные параметры
        this.useCofactor = useCofactor; this.keyWrapParameters = keyWrapParameters; 
    } 
    // конструктор
    public KeyAgreement(boolean useCofactor) 
    { 
        // сохранить переданные параметры
        this.keyDerive = new NOKDF(ENDIAN); 
            
        // сохранить переданные параметры
        this.useCofactor = useCofactor; this.keyWrapParameters = null;   
    } 
    // освободить используемые ресурсы
    @Override protected void onClose() throws IOException 
    {
        // освободить используемые ресурсы
        RefObject.release(keyDerive); super.onClose();            
    }
    // сгенерировать случайные данные
    @Override public byte[] generate(IParameters parameters, IRand rand) throws IOException
    {
        // проверить необходимость генерации
        if (keyDerive instanceof NOKDF) return null; 
        
        // сгенерировать случайные данные
        byte[] random = new byte[64]; rand.generate(random, 0, 64); return random;   
    }
    // наследовать ключ
    @Override public ISecretKey deriveKey(IPrivateKey privateKey, 
        IPublicKey publicKey, byte[] random, 
        SecretKeyFactory keyFactory, int keySize) throws IOException
    {
        // преобразовать тип данных
        aladdin.capi.ansi.x962.IPrivateKey ecPrivateKey = 
            (aladdin.capi.ansi.x962.IPrivateKey)privateKey; 
        aladdin.capi.ansi.x962.IPublicKey  ecPublicKey  = 
            (aladdin.capi.ansi.x962.IPublicKey )publicKey; 

        // получить параметры алгоритма
        aladdin.capi.ansi.x962.IParameters ecParameters = 
            (aladdin.capi.ansi.x962.IParameters)privateKey.parameters(); 
            
        // получить параметры эллиптической кривой
        aladdin.capi.ec.Curve ec = ecParameters.getCurve(); 
            
        // вычислить точку на эллиптической кривой
        ECPoint P = ec.multiply(ecPublicKey.getW(), ecPrivateKey.getS()); 
            
        // выполнить дополнительное умножение
        if (useCofactor) P = ec.multiply(P, BigInteger.valueOf(ecParameters.getCofactor()));
        
        // при наличии параметров шифрования
        if (keyWrapParameters != null)
        {
            // закодировать случайные данные
            OctetString entityUInfo = (random != null) ? new OctetString(random) : null; 
        
            // закодировать размер ключа в битах
            OctetString suppPubInfo = new OctetString(Convert.fromInt32(keySize * 8, ENDIAN));
            
            // объединить закодированные данные
            SharedInfo sharedInfo = new SharedInfo(
                keyWrapParameters, entityUInfo, null, suppPubInfo, null
            ); 
            // получить закодированное представление
            random = sharedInfo.encoded(); 
        }
        // определить размер закодированного представления
        int cb = (ec.getField().getFieldSize() + 7) / 8; 
        
        // закодировать координату точки
        byte[] encodedX = Convert.fromBigInteger(P.getAffineX(), ENDIAN, cb); 
        
        // закодировать координату точки
        try (ISecretKey z = keyDerive.keyFactory().create(encodedX))
        {
            // выполнить наследование ключа
            return keyDerive.deriveKey(z, random, keyFactory, keySize); 
        }
        // обработать неожидаемое исключение
        catch (InvalidKeyException e) { throw new RuntimeException(e); }
    }
}

File: Source/Aladdin.CAPI.ANSI/CAPI/ANSI/Keyx/ECDH/KeyAgreement.cs
using System;

namespace Aladdin.CAPI.ANSI.Keyx.ECDH
{
    ///////////////////////////////////////////////////////////////////////////
    // Формирование общего ключа Elliptic Curve Diffie-Hellman
    ///////////////////////////////////////////////////////////////////////////
    public class KeyAgreement : CAPI.KeyAgreement
    {
        // способ кодирования чисел
        private const Math.Endian Endian = Math.Endian.BigEndian; 

        // алгоритм наследования ключа
        private KeyDerive keyDerive; private bool useCofactor; 
        // параметры алгоритма шифрования ключа
        private ASN1.ISO.AlgorithmIdentifier keyWrapParameters; 
    
        // конструктор
        public KeyAgreement(bool useCofactor, KeyDerive keyDerive, 
            ASN1.ISO.AlgorithmIdentifier keyWrapParameters)
        { 
            // сохранить переданные параметры
            this.keyDerive = RefObject.AddRef(keyDerive); 
        
            // сохранить переданные параметры
            this.useCofactor = useCofactor; this.keyWrapParameters = keyWrapParameters; 
        } 
        // конструктор
        public KeyAgreement(bool useCofactor, CAPI.Hash hashAlgorithm, 
            ASN1.ISO.AlgorithmIdentifier keyWrapParameters)
        { 
            // сохранить переданные параметры
            this.keyDerive = new Derive.X963KDF(hashAlgorithm); 
            
            // сохранить переданные параметры
            this.useCofactor = useCofactor; this.keyWrapParameters = keyWrapParameters; 
        } 
        // конструктор
        public KeyAgreement(bool useCofactor)
        { 
            // сохранить переданные параметры
            this.keyDerive = new CAPI.Derive.NOKDF(Endian); 
        
            // сохранить переданные параметры
            this.useCofactor = useCofactor; this.keyWrapParameters = null; 
        } 
        // освободить используемые ресурсы
        protected override void OnDispose() 
        {
            // освободить используемые ресурсы
            RefObject.Release(keyDerive); base.OnDispose();
        }
        // сгенерировать случайные данные
        public override byte[] Generate(IParameters parameters, IRand rand)
        {
            // проверить необходимость генерации
            if (keyDerive is CAPI.Derive.NOKDF) return null; 

            // сгенерировать случайные данные
            byte[] random = new byte[64]; rand.Generate(random, 0, 64); return random;   
        }
        // наследовать ключ
        public override ISecretKey DeriveKey(IPrivateKey privateKey, 
            IPublicKey publicKey, byte[] random, SecretKeyFactory keyFactory, int keySize)
        {
            // преобразовать тип данных
            X962.IPrivateKey ecPrivateKey = (X962.IPrivateKey)privateKey; 
            X962.IPublicKey  ecPublicKey  = (X962.IPublicKey )publicKey; 

            // получить параметры алгоритма
            X962.IParameters ecParameters = (X962.IParameters)privateKey.Parameters; 
        
            // получить параметры эллиптической кривой
            EC.Curve ec = ecParameters.Curve; 
        
            // вычислить точку на эллиптической кривой
            EC.Point P = ec.Multiply(ecPublicKey.Q, ecPrivateKey.D); 
        
            // выполнить дополнительное умножение
            if (useCofactor) P = ec.Multiply(P, ecParameters.Cofactor); 

            // при наличии параметров шифрования
            if (keyWrapParameters != null)
            {
                // закодировать случайные данные
                ASN1.OctetString entityUInfo = (random != null) ? new ASN1.OctetString(random) : null; 

                // закодировать размер ключа в битах
                ASN1.OctetString suppPubInfo = new ASN1.OctetString(
                    Math.Convert.FromInt32(keySize * 8, Endian)
                );
                // объединить закодированные данные
                ASN1.ANSI.X962.SharedInfo sharedInfo = new ASN1.ANSI.X962.SharedInfo(
                    keyWrapParameters, entityUInfo, null, suppPubInfo, null
                ); 
                // получить закодированное представление
                random = sharedInfo.Encoded; 
            }
            // определить размер закодированного представления
            int cb = (ec.Field.FieldSize + 7) / 8; 
        
            // закодировать координату точки
            byte[] encodedX = Math.Convert.FromBigInteger(P.X, Endian, cb); 

            // закодировать координату точки
            using (ISecretKey z = keyDerive.KeyFactory.Create(encodedX))
            {
                // выполнить наследование ключа
                return keyDerive.DeriveKey(z, random, keyFactory, keySize); 
            }
        }
    }
}
